rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,487 @@
+most importantly, the system core deserves the most attention, as it is the
+central component that maintains the internal state of VAST.
+
+From a fault-tolerance perspective, the system core is the *catalog*, as it owns
+all data and metadata. VAST stores data into partitions. Mutating operations
+like compaction, rollup, or deletion of partitions must leave the catalog in an
+internally consistent state. The catalog provides atomic functions for
+insertion, replacement and removal of partitions to support this requirement.
+The current implementation of the catalog is based on a custom in-memory store,
+without support for distribution or replication. It is therefore not suitable
+for environments with stronger fault tolerance requirements.
+
+## Requirements
+
+We can frame the requirements we are facing more generally in terms of the CAP
+Theorem. Given that availability is of utmost importance for processing security
+data, we cannot compromise on it. On the contrary, we can compromise on
+consistency, given that telemetry data is fundamentally incomplete. Security
+analysts are always working with a subset of descriptions of activity, and
+stitching together analytics based on partial knowledge is the norm. For this
+reason, it makes sense to also prioritize availability when facing a network
+partition.
+
+Unlike a transactional RDBMS, the system expectations for our use case center
+around *eventual consistency*. For example, it is acceptable for the system to
+have slightly different knowledge about existing data, and thereby produce
+slightly different query results in the short term, as long these discrepancies
+consolidate eventually.
+
+With respect to the system core, this implies that the catalog must support
+transactions when processing partition operations. For example, a compaction
+operation may remove N partitions and replace them with a new one. These
+operations break down to a sequence of deletes, followed by an insert.
+
+### Non-requirements
+
+Infrastructure approaches for hardware redundancy are considered out of scope.
+In particular, we do not consider fault tolerance at the storage layer. There
+are many different ways to provide redundant storage, e.g., RAID storage,
+distributed file systems or NAS. Cloud storage, like S3 or Azure blob storage,
+are replicated three times, and adding an additional layer of fault tolerance on
+top does not provide a lot of additional safety but considerably increases the
+cost.
+
+## Solution Proposal
+
+The most important architectural effort concerns making the catalog a
+fault-tolerant component. To this end, we propose to factor the logic that
+operates on partition metadata into a dedicated **metastore** component. As a
+first point of contact for queries, the catalog also keeps per-partition index
+structures in memory. This catalog functionality is out of scope, as it
+represents derived, immutable state from the metastore.
```

Comment:
I reserve myself the freedom use the word informally here. These questions can only be answered with a lot more context, meaning further down the document.

For you: Whether the metastore is encapsulated in the catalog or exposed as a standalone service actor depends on whether we want to make it available to other clients besides the catalog. It might also depend on external processes services depending on the backing implementation. How this is all organized is an exercise in code organization.

Reply:
The term "component" is already established in VAST, and everyone knowing that will read this sentence differently (as did I). That's why I'm asking you to make this more explicit.

Reply:
"component" is still the most suitable term and it is used intentionally here. We should not call it an actor because several of the potential backends would run in a separate process that would only be wrapped by an actor, others would consist of multiple actors.
I will also refrain from specifying who shall create / own the metastore here. Artificial constraints only add complexity to the overall design.

Reply:
We talked about this and found a compromise: Just take the work "component" out, the sentence still works the same.